core/modules/replica/MessengerConnector.h

```diff
@@ -0,0 +1,523 @@
+/*
+ * LSST Data Management System
+ * Copyright 2017 LSST Corporation.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+#ifndef LSST_QSERV_REPLICA_MESSENGERCONNECTOR_H
+#define LSST_QSERV_REPLICA_MESSENGERCONNECTOR_H
+
+/// MessengerConnector.h declares:
+///
+/// class MessengerConnector
+/// (see individual class documentation for more information)
+
+// System headers
+#include <functional>
+#include <list>
+#include <map>
+#include <memory>
+#include <stdexcept>
+#include <string>
+
+// Third party headers
+#include <boost/asio.hpp>
+
+// Qserv headers
+#include "proto/replication.pb.h"
+#include "replica/ProtocolBuffer.h"
+#include "replica/ServiceProvider.h"
+#include "util/Mutex.h"
+
+// This header declarations
+
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+// Forward declarations
+class WorkerInfo;
+
+/**
+ * Class MessengerConnector provides a communication interface for sending/receiving
+ * messages to and from worker services. It provides connection multiplexing and
+ * automatic reconnects.
+ *
+ * NOTES ON THREAD SAFETY:
+ *
+ * - in the implementation of this class a mutex is used to prevent race conditions
+ *   when performing internal state transitions.
+ *
+ * - to avoid deadlocks, only externally called methods of the public API (such
+ *   as the ones for sending or cancelling requests) and assynchronious callbacks
+ *   are locking on the mutex. Those methods are NOT allowed to call each other.
+ *   Otherwise deadlocks are imminent.
+ *
+ * - other implementatin-specific methods are supposed to be invoked from either
+ *   of the above mentioned methods.
+ */
```

Comment:
it's not totally clear what you mean by "other implementation-specific methods". I think you're saying that public api functions take a lock, and do not call each other. And protected/private functions expect that a lock is already taken and may call each other (but not may not call public functions). Is that right? (Maybe change the wording to make it more clear?)

Reply:
That's right. I'm fixing this line of the class's documentation.